---
title: Epigenetics responses to excess heat and CO$_{2}$ selection in the copepod _Acartia tonsa_
author: \emph{Thomas O'Leary}
output:
  pdf_document:
header-includes:
  - \usepackage[cmintegrals,cmbraces]{newtxmath}
  - \usepackage{ebgaramond}
  - \usepackage[T1]{fontenc}
  - \usepackage{graphicx}
  - \usepackage{float}
  - \usepackage{wrapfig}
  - \usepackage[font={footnotesize}]{caption}
bibliography: tso_refs.bib
csl: mla.csl
fontsize: 11pt
geometry: "left=1cm,right=1cm,top=1cm,bottom=1cm"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE)
require(tidyverse)
require(kableExtra)

# Load the epigenetic data -----------------------------------------------------
source(here::here("myscripts/load_epigenetics.R"))
```

# Background

### _Acartia tonsa_ introduction

A world wide distribution and a huge population size make the copepod, _Acartia tonsa_, a particularly interesting species to study in relation to epigenetic changes in response to the environment. _A. tonsa_ lives and thrives in a broad range of habitats, including large variation in temperature, salinity, and pH [@Chen2011]. Expected increases in sea surface temperature and decreases in pH (from absorbed atmospheric carbon dioxide) driven by climate change, make _A. tonsa_ a particularly interesting speices to study in the context of heat and carbon dioxide stress. In this experiment, natural collections of _A. tonsa_ were subjected to 25 generations of selection under heat and carbon dioxide stress treatments. Methylation of the genome was quantified before and after experimental treatment. Here, I will limit my focus to the effect of heat stress and combined heat and carbon dioxide stress on _A. tonsa_ after 25 generations of selection. The aims of this study are to (i) describe the epigenetic response to heat stress and (ii) describe how the additional stressor of carbon dioxide affects that response, and (iii) to identify the genes associated with the loci that are differentially methylated.

### Selection experiment, sample collection, library preparation, and sequencing

Natural collections of _Acartia tonsa_ were kept in a common garden for three generations, before being split for 25 generations in four separate treatment groups: (i) a control with ambient temperature (18°C) and ambient CO$_{2}$ levels (400 ppm), (ii) a high temperture treatment (22°C) with ambient CO$_{2}$ levels, (iii) a high CO$_{2}$ (2000 ppm) treatment with ambient temperature and, (iv) a combined high temperature and high CO$_{2}$ treatment. There were four replicates of each treatment, with approximately 3,000--5,000 individuals per replicate. Whole genomic \textsc{DNA} was extracted from pools of ~30 individuals and collected at generation zero and generation 25. Reduced representation bisulfite sequencing (RRBS) library preparation was used to quantify methlyation frequencies, where all unmethylated cytosines are converted to thymines. The samples were 150 bp sequenced on an Illumina HiSeq X.

	
# Bioinformatics Pipeline

The quality of the raw reads were accessed using FastQC v$0.11.7$ [@andrews2012] and low quality sequence data was trimmed using Trimmomatic v$0.33$ [@Bolger2014]. The cleaned reads were vizualized again with FastQC and then mapped to the _A. tonsa_ reference genome [@Jørgensen2019] with Bismark v$0.22.1$ [@Krueger2011]. The reference genome needed to be modified from its original version to reflect the RRBS treatment into two versions: one with the cytosines converted to thymines for the forward reads (Read 1) and the other with guanines converted to adenines for the reverse reads (Read 2). Differential methylation was called at and false discovery rate of $0.05$ and a minimum relative fold-change of $10$% [^sig] using the package methylKit v$1.13.1$ [@Akalin2012]  R v$3.6.1$ [@rcore]. The methylation data was pre-filtered to include only loci that had at least 20 reads and to exculed loci in the highest $2.5$% of per base read coverage[^filt]. Other analysis and visualization was carried out in R using ggplot2 and dplyr v$1.2.1$ [@tidyverse].

[^sig]: We required a minimum relative fold-change of $10$% for differential methylation in order to exclude statistically significant, but minor relative fold-changes that likely have little-to-no biological significance.

[^filt]: The pre-filtering was done to ensure that there was enough data to allow for an accurate estimate of the percent methylation (low-end limit) and remove data that might be the product of procedural sequencing error (high-end limit).

# Results

The average mapping rate across all samples was $45.2$%, with averages for the subset of groups I analyzed of $43.2$%, $33.9$%, and $31.0$% for the generation 25 control, high temperature only, and combined high temperature and high carbon dioxide groups respectively. The global methylation rate was not drastically different between the three treatment groups after 25 generations, with a mean per site methylation rate of all three groups between $38$% and $39$% (Fig. 1 **A**). However, there were still individual loci that showed differential methylation between the experimental treatments and the control. We identified 105 differentially methylated loci (DML) in response to heat stress, with 84 hypermethylated loci and 21 hypomethylated loci (Fig. 1 **B**). Under combined heat stress and excess CO$_2$, 121 DML were idenified, with 66 hypermethylated loci and 55 hypomethylated loci (Fig. 1 **B**). There was no overlap between the lists of DML. Those DML mapped to 9 unique differentially methylated genes (DMGs) for the heat stress and 13 DMGs for combined the heat and CO$_2$ stress, of which 5 DMGs were overlapping between the groups. 

```{r, fig.width = 8, fig.height = 3, fig.cap = paste0("\\textbf{(A)} Mean per site methylation rate (%) across all sites included in the analysis. \\textbf{(B)} Frequency histogram of the percent difference in methylation of only the differenitally methylated loci (DML) for each treatment, Heat (pink) and Heat & CO$_2$ (dark red), compared to the control.")}
p1 <- ggplot(pm, aes(x = Group, y = avg_meth, fill = Group)) +
  stat_summary(color = "grey50", fun.data = "mean_se") +
  geom_jitter(width = 0.1, size = 3, pch = 21, alpha = 0.8) + 
  ylab("Global methylation rate (%)") +
  xlab("Treatment") +
  ylim(c(35, 42)) +
  scale_fill_manual(values = c("grey50", "pink", "darkred")) +
  scale_x_discrete(limits = c("AA", "HA", "HH"),
                   labels = c("Control", 
                              "Heat", 
                              expression("Heat & CO"[2]))) +
  theme_classic() +
  theme(legend.position = "none")


p2 <- ggplot(hist.plot, aes(x = meth.diff, fill = Treatment)) +
  geom_histogram(position = "dodge", color = "grey50", bins = 20) +
  geom_vline(xintercept = 0) +
  labs(x = "Methylation difference to control",
       y = "Frequency") + 
  theme_classic() + 
  scale_fill_manual(values = c("pink", "darkred"),
                    labels = c("Heat", expression("Heat & CO"[2]))) +
  theme(legend.position = c(0.2, 0.85), 
        legend.background = element_rect(linetype = 1, color = 1),
        legend.text.align = 0)

cowplot::plot_grid(p1, p2, nrow = 1, labels = "AUTO", rel_widths = c(1, 1.5))
```

One of the DMGs 

# Conclusion

The total number of identified DML was only slightly higher in the combined Heat & CO$_2$ treatment, but there is a significant difference in the distribution of the two treatments (Kolmogorov-Smirnov tests: p = `r signif(KS_pval, 2)`), with the heat stress only treatment having disproportionately more hypermethylated loci compared to the relatively even distribution in the combined heat & excess CO$_2$ treatment (Fig 1. **B**).

The gene ontology of the identified DMGs indicate that these genes are involved in a variety of biological processes. Three of the five DMGs that are shared between the two treatments, Retrovirus-related Pol polyprotein from transposon opus, Gypsy retrotransposon integrase-like protein 1, and Endogenous retrovirus group K member 5 Gag polyprotein are involved in biological processes such as DNA integration, recombination, and RNA mediated transposition. This could refelct a trend of differential methylation of genes that cotribute to increasing genetic diversity in the population through recombination and transposition.


It should be noted that here we identify DMGs only with respect to the pairwise comparisons of high temperature and combined high temperature and high CO$_2$ to the control treatment. This approach does not include the full model of treatments and time points. Pairwise comparisons used in this way, ignoring the full experimental design, may lead to an increase in false positives, so care must be taken when interpreting the lists of DML. One further possible complication of this analysis is the fact that this data was taken from serveral pooled whole-body individuals, leading to greater noise and potentially masking important individual and tissue type epigenetic variation. 

Future directions should involve incorporating transcriptomic and genomic sequencing data with this epigenetic data. Despite finding several DMGs, it is not know how that may or may not affect the expression of a gene without the corresponding transcriptomic data set. Along the same lines, the epigenetic and transcriptomic data together do not tell the full story either. Adaptive allele frequency changes in regulatory and coding regions may result in differential gene expression. To understand if their really has been adaptive change in these experimental pools of _A. tonsa_, you must measure certain fitness related phenotpyes, like survival, fecundity, and growth. And do a reciprocal transplant to see if the control pools fair worse under those same environmental stressors.



\pagebreak

# References
